consensus/polybft/proposer_calculator.go

```diff
@@ -0,0 +1,356 @@
+	mostest, err := pc.getValWithMostPriority()
+	if err != nil {
+		return nil, fmt.Errorf("cannot get validator with most priority: %w", err)
+	}
+
+	mostest.ProposerPriority = safeSubClip(mostest.ProposerPriority, pc.totalVotingPower)
+
+	return mostest, nil
+}
+
+func (pc *proposerCalculator) updateWithChangeSet() error {
+	if err := pc.rescalePriorities(); err != nil {
+		return fmt.Errorf("cannot rescale priorities: %w", err)
+	}
+
+	if err := pc.shiftByAvgProposerPriority(); err != nil {
+		return fmt.Errorf("cannot shift proposer priorities: %w", err)
+	}
+
+	return nil
+}
+
+func (pc *proposerCalculator) shiftByAvgProposerPriority() error {
+	avgProposerPriority, err := pc.computeAvgProposerPriority()
+	if err != nil {
+		return fmt.Errorf("cannot compute proposer priority: %w", err)
+	}
+
+	for _, val := range pc.validators {
+		val.ProposerPriority = safeSubClip(val.ProposerPriority, avgProposerPriority)
+	}
+
+	return nil
+}
+
+func (pc *proposerCalculator) getValWithMostPriority() (result *validatorCalcMetadata, err error) {
+	if pc.isNilOrEmpty() {
+		return nil, fmt.Errorf("validators cannot be nil or empty")
+	}
+
+	for _, curr := range pc.validators {
+		// pick curr as result if it has greater priority
+		// or if it has same priority but "smaller" address
+		if curr.IsBetter(result) {
+			result = curr
+		}
+	}
+
+	return result, nil
+}
+
+func (pc *proposerCalculator) computeAvgProposerPriority() (int64, error) {
+	if pc.isNilOrEmpty() {
+		return 0, fmt.Errorf("validator set cannot be nul or empty")
+	}
+
+	n := int64(len(pc.validators))
+	sum := big.NewInt(0)
+
+	for _, val := range pc.validators {
+		sum.Add(sum, big.NewInt(val.ProposerPriority))
+	}
+
+	avg := sum.Div(sum, big.NewInt(n))
+	if avg.IsInt64() {
+		return avg.Int64(), nil
+	}
+
+	return 0, fmt.Errorf("cannot represent avg ProposerPriority as an int64 %v", avg)
+}
+
+// rescalePriorities rescales the priorities such that the distance between the
+// maximum and minimum is smaller than `diffMax`.
+func (pc *proposerCalculator) rescalePriorities() error {
+	if pc.isNilOrEmpty() {
+		return fmt.Errorf("validator set cannot be nul or empty")
+	}
+
+	// Cap the difference between priorities to be proportional to 2*totalPower by
+	// re-normalizing priorities, i.e., rescale all priorities by multiplying with:
+	// 2*totalVotingPower/(maxPriority - minPriority)
+	diffMax := priorityWindowSizeFactor * pc.totalVotingPower
+
+	// Calculating ceil(diff/diffMax):
+	// Re-normalization is performed by dividing by an integer for simplicity.
+	// NOTE: This may make debugging priority issues easier as well.
+	diff := computeMaxMinPriorityDiff(pc.validators)
+	ratio := (diff + diffMax - 1) / diffMax
+
+	if diff > diffMax {
+		for _, val := range pc.validators {
+			val.ProposerPriority /= ratio
+		}
+	}
+
+	return nil
+}
+
+// isNilOrEmpty returns true if validator set is nil or empty.
+func (pc proposerCalculator) isNilOrEmpty() bool {
+	return len(pc.validators) == 0
+}
```

Comment:
What do you think about inlining this function? It is a one-liner and only checks single thing. If it is going to remain, then I think it should be renamed to e.g. `isEmptyValidatorSet`.

Reply:
Removed https://github.com/0xPolygon/polygon-edge/pull/984/commits/a3e993bda514d24947454feaf47d95846ce64edd